src/stateMatrixParser.cpp

```diff
@@ -0,0 +1,443 @@
+    while (it != s.end() && std::isdigit(*it)) ++it;
+    return !s.empty() && it == s.end();
+}
+
+std::vector<unsigned int> recreate_tm(StateMatrix stM, unsigned int states, unsigned int  alphabet_size, unsigned int numIt){
+
+  TuringMachine machine(states, alphabet_size);    
+  machine.set_state_matrix(stM); 
+  machine.reset_tape_and_state();
+  
+  for (auto i = 0u; i < numIt -1 ; ++i){
+    machine.act();
+  }
+
+  return machine.get_written_tape();
+}
+
+std::pair<std::vector<unsigned int>,StateMatrix> random_tm(Rng rng, unsigned int states, unsigned int  alphabet_size, unsigned int numIt){
+  TuringMachine machine(states, alphabet_size);    
+  machine.stMatrix.set_random(rng);
+  machine.reset_tape_and_state();
+  
+  for (auto i = 0u; i < numIt -1 ; ++i){
+    machine.act();
+  }
+  std::pair<std::vector<unsigned int>,StateMatrix> pair_tape_state_matrix(machine.get_written_tape(),machine.get_state_matrix());
+  return pair_tape_state_matrix;
+}
+
+std::pair<std::vector<double>,StateMatrix> mutate_rules_once_at_a_time(StateMatrix stM, unsigned int states, unsigned int  alphabet_size, unsigned int numIt,BestKMarkovTables<NormalizedCompressionMarkovTable>& bestMkvTableCompression){
+  bool baseline = true;
+  std::vector<Metrics> all_other_permutation_metrics_in_rules;
+  std::vector<Metrics> all_other_permutation_metrics_avg_per_rule;
+  Metrics baselineTM;
+  
+  for (auto index = 0u; index<stM.get_size(); index++){
+    
+    TuringRecord tr{0,0,0};
+    TuringRecord element_tr = stM.get_element(index);
+    
+    if(!(tr==element_tr)){
+      stM.set_rule(index,tr);
+      auto vector_written_string = recreate_tm(stM, states, alphabet_size, numIt);
+      all_other_permutation_metrics_in_rules.push_back(bestMkvTableCompression.update_string(vector_written_string));
+      bestMkvTableCompression.reset();
+    }
+    else if(baseline){
+      auto vector_written_string = recreate_tm(stM, states, alphabet_size, numIt);
+      baselineTM = bestMkvTableCompression.update_string(vector_written_string);
+      bestMkvTableCompression.reset();
+      baseline = false;
+    }
+      
+    while(tr.next(states,alphabet_size)){
+      if(!(tr==element_tr)){
+        stM.set_rule(index,tr);
+        auto vector_written_string = recreate_tm(stM, states, alphabet_size, numIt);
+        all_other_permutation_metrics_in_rules.push_back(bestMkvTableCompression.update_string(vector_written_string));
+        bestMkvTableCompression.reset();
+      }
+      else if(baseline){
+      auto vector_written_string = recreate_tm(stM, states, alphabet_size, numIt);
+      baselineTM= bestMkvTableCompression.update_string(vector_written_string);
+      bestMkvTableCompression.reset();
+      baseline = false;
+      }
+    }
+    all_other_permutation_metrics_avg_per_rule.push_back(avg_metrics(all_other_permutation_metrics_in_rules));
+    stM.set_rule(index,element_tr);
+  }
+  std::vector<double> rule_relevance_vector;
+  double rule_relevance;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The scope of the variable 'rule_relevance' can be reduced.](https://app.codacy.com/app/jorgeMFS/TMCompression/pullRequest?prid=4494377)